File: Week04/LeetCode-0001.cpp
class Solution {
public:
    /// O(N) algorithm with O(N) extra space
    vector<int> twoSum(vector<int>& nums, int target) {
        //           <num, index>
        unordered_map<int, int> myMap;

#if 0        
        for (int i=0; i<nums.size(); i++){
            int otherNum = target-nums[i];
            auto iter = myMap.find(otherNum);
            if (iter != myMap.end())
                return {iter->second, i};
            else
                myMap.insert({nums[i], i});
        } // end-for
#else
        for (int i=0; i<nums.size(); i++){
            int otherNum = target-nums[i];
            if (myMap.count(otherNum) != 0)
                return {myMap[otherNum], i};
            else
                myMap[nums[i]] = i;
        } // end-for        
#endif        
        return {-1, -1};
    } // end-twoSum
};

File: Week04/LeetCode-0001.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        O(N) algorithm with O(N) extra space
        """
        myMap = dict()
        
        for i in range(len(nums)):
            otherNum = target-nums[i]
            if otherNum in myMap:
                return {myMap[otherNum], i}
            else:
                myMap[nums[i]] = i
        return {-1, -1}

        